src/cpu_shader/euler.rs

```diff
@@ -0,0 +1,277 @@
+// Copyright 2023 the Vello Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT OR Unlicense
+
+//! Utility functions for Euler Spiral based stroke expansion.
+
+// Use the same constants as the f64 version.
+#![allow(clippy::excessive_precision)]
+
+use super::util::Vec2;
+use std::f32::consts::FRAC_PI_4;
+
+#[derive(Debug)]
+pub struct CubicParams {
+    pub th0: f32,
+    pub th1: f32,
+    pub d0: f32,
+    pub d1: f32,
+}
+
+#[derive(Debug)]
+pub struct EulerParams {
+    pub th0: f32,
+    pub th1: f32,
+    pub k0: f32,
+    pub k1: f32,
+    pub ch: f32,
+}
+
+#[derive(Debug)]
+pub struct EulerSeg {
+    pub p0: Vec2,
+    pub p1: Vec2,
+    pub params: EulerParams,
+}
+
+impl CubicParams {
+    /// Compute parameters from endpoints and derivatives.
+    pub fn from_points_derivs(p0: Vec2, p1: Vec2, q0: Vec2, q1: Vec2, dt: f32) -> Self {
+        let chord = p1 - p0;
+        // Robustness note: we must protect this function from being called when the
+        // chord length is (near-)zero.
```

Comment:
Maybe the CPU shaders should assert that the chord length is less than some epsilon (I'm not sure if Rust panics on a floating point division by zero).

Reply:
It does not. I'll add an assert and some comments, that should help us debug the case.

And fun! When I added the assertion it got triggered. I'll add the robustness logic for this now.